### ISSUE TITLE
Scoping of definitions

### ISSUE THREAD

**Issue by hddharvey** [skvadrik/re2c#366]
This is just an observation that I've made while learning and using re2c. Feel free to use these to improve it (or not - it's pretty minor).

Definition scope is global across all re2c blocks. This can be useful but I've found it mostly to be slightly annoying when working with multiple lexers in the same file since I have to be careful to use different definition names - which can artificially make things more verbose. If I do want to share definitions between re2c blocks, it's also not very readable since it breaks the programmer's normal intuitions about scoping - a re2c block feels like a "scope", so it's almost as if you did this in C++:
```cpp
{
    int x = 0;
}
// ...
{
    x += 5; // this would be a compiler error in C++
}
```

This also seems to go against how re2c normally treats these blocks. Having separate /* */ blocks generates separate lexers, so intuitively feels like a separate "invocation" of re2c at each point. You could argue that things like separate "rules" blocks or "include" blocks contradict that, but these feel like explicitly global declarations since it's in the name of the block itself.

I wasn't able to find any ways in the manual to control the scoping of definitions/etc. to be 'local' or 'global'.

I understand it may be not worthwhile or not feasible to implement this (since it would be a breaking change, so you would either control it with a re2c flag or provide a way to enable it in the .re file - e.g., a way to mark re2c blocks as 'local' or 'scoped', or a global configuration option for a .re file).

If you were to control it with a re2c flag, ideally the program would not compile with the flag set incorrectly, or if it did compile with and without the flag, the program would have the same meaning in both cases. I think this suggestion satisfies satisfies that, since reusing a definition in a further-down block would be an error with blocks set to be 'scoped', and redefining a name in a new block would be a redefinition error with blocks set to 'unscoped'.

Anyway, pretty minor, but I thought I'd raise it.

**Comment by skvadrik** [comment]
It's a good idea, and I have recently added support for it: named blocks and use directive that allows one to include a whole block inside of another block. You can see a full example here: https://github.com/skvadrik/re2c/blob/master/examples/c/reuse/usedir.re. I have also updated manpages (https://github.com/skvadrik/re2c/blob/master/bootstrap/doc/re2c.1#L2792-L2821) but not the online docs, as the feature is not released yet.

Basically, what you do is:
- declare a named block `/*!rules:re2c:<block-name-1> ... */` with some definitions inside
- declare another named block `/*!rules:re2c:<block-name-2> ... */` with some other definitions inside
- have a lexer block `/*!re2c ... */` and use the named block `<block-name-1>` with the help of the new use directive: `!use:<block_name-1>;` (to be used inside of the lexer block)
- have another lexer block that uses the second one: `use:<block-name-2>;`

And so on, you can combine named blocks freely with the help of use directive. Replace the block names in angles with some real names. No new option is needed --- alhough the new syntax was previously allowed, it would parse as a syntax error because a colon followed by a block name would be interpreted as the beginning of code inside of a block (and that is ill-formed code).

It would be great if you could try this new functionality in your projects and give some feedback if you encounter difficulties or limitations --- it is still possible to tweak things before releasing this feature and updating the online docs.

**Comment by skvadrik** [comment]
To clarify, named blocks can have rules and configurations as well --- they are like ordinary blocks.

Also, `-r` option that enables reusable blocks is no longer needed --- it will be on by default in the new release.

**Comment by hddharvey** [comment]
When a normal `!re2c` block `!use`s a `!rules` block, do the same scoping rules apply? I.e., are the definitions imported by the rules block visible in other `!re2c` blocks further down?

I'll see if I can try it out later.

**Comment by skvadrik** [comment]
> When a normal !re2c block !uses a !rules block, do the same scoping rules apply? I.e., are the definitions imported by the rules block visible in other !re2c blocks further down?

Yes, for the normal blocks definitions are still accumulated. E.g. in this example `x` definition is visible in the last block:
```
/*!rules:re2c:x
    x = [a];
*/

/*!re2c
    !use:x;
*/

/*!re2c
    x {}
*/
```

I see what you mean, you'd like a way to make the `x` definition to be scoped to the `x` block and the one that uses it. This indeed would require new syntax. I see two options:

1. a new type of block (someting like `/*!local:re2c ... */`)
2. a new option and configuration (e.g. `--local-scope` and `re2c:local-scope = 1;`)

My preference would be 1, because it requires less boilerplate from the programmer, it allows to have local and global blocks in the same program (unlike the global option), and older versions of re2c would parse it as a comment (which would result in not generating any code and no affecting the global scope). The downside of a global option is that it is unlikely that all re2c users  would ever migrate to it, so it will never be the default.

**Comment by hddharvey** [comment]
I agree that 1 would be preferable.

You can't make local block the default without introducing a breaking change. However, making local blocks the default will only make some re2c programs invalid and leave all others unaffected (since it is not possible to shadow definitions) - so it's not _that_ bad of a breaking change. The worst kind of breaking change is one that silently changes program behavior. In cases such as this though, it would be easy to issue an informative diagnostic for undeclared symbols in newly broken programs by checking if that system would have been declared under global scoping and telling the user how to fix it (by opting in to global scoping or using a rules block to share definitions, or whatever).

For me, I don't mind what approach is taken. I would be happy to write "!local" at the start of blocks. However, in an ideal world, I think a more scope-based approach makes more intuitive sense and would be a better default, but I understand that breaking changes are not desirable and an option like 1 or 2 might be preferable.

**Comment by skvadrik** [comment]
Thanks for your thoughtful feedback @hddharvey.

The next release (currently in development) will add support for named blocks. It will be possible to specify names for all blocks, not only `/*!rules:re2c... */` and `/*!use:re2c... */`, but also for the usual code blocks. This is convenient because these names can be later referenced in directives such as `/*!stags:re2c... */`, `/*!getstate:re2c... */` or `/*!max:re2c... */` that previously accumulated data across the whole file. Now it will be possible to have block-granularity directives. I haven't fully implemented it yet, so it is subject to change, but I'm working on it now.

For the named blocks, since they are new and not bound by backwards compatibility, we can set local scope by default. The unnamed blocks will continue to be global. This seems rather natural: empty namespace is the global namespace. Then we can rewrite tutorials to put more emphasis on the named blocks and perhaps eventually add a warning if unnamed blocks are used (but that would be rather annoying).

> P.S.: if it wasn't clear, my original comment was referring to any kind of named declaration, whether it be definitions or rules. 

Yes, named blocks add scoping for definitions, configurations and rules.

**Comment by skvadrik** [comment]
> For the named blocks, since they are new and not bound by backwards compatibility, we can set local scope by default.

@trofi pointed out that it might be a bad idea to flip defaults between named and unnamed blocks, because someone might try to convert a program with unnamed blocks into a program with named blocks and have subtle issues with configurations not being applied anymore, like in this example:
```
/*!re2c
    re2c:some:option = 1; // set in global scope
*/

/*!re2c
    // ... code that expects 'some:option' to be set
*/
```
After converting to named blocks, assuming that named blocks have local scope by default:
```
/*!re2c:someblock
    re2c:some:option = 1; // set in local scope
*/

/*!re2c
    // ... 'some:option' is not set
*/
```
Depending on a particular option this can cause a re2c error (ok, easy to notice and fix), compiler error (also explicit for the user) or a silent change of behavior (too bad and should be avoided).

So it is perhaps better to stick to `/*!local:re2c ... */` syntax.

**Comment by skvadrik** [comment]
Support for local blocks has been merged to master. Here is the commit that adds it: https://github.com/skvadrik/re2c/commit/15cc78861a7faffc89087f2756ed04523e16b902

**Comment by hddharvey** [comment]
@skvadrik Local blocks works for me. Thanks for implementing this.
